### PR TITLE
Add torchvision maintainers guide

### PR DESCRIPTION
**Pull request by NicolasHug**
This guide properly documents our deprecation policy (which was introduced a while ago) and establishes the distinction between public and private APIs - following the torch core policy.

[Rendered version](https://github.com/NicolasHug/vision/blob/maintainersguide2/maintainer_guide.md)

CC @pmeier

### CODE REVIEW

File: maintainer_guide.md
Comment (pmeier) [inline maintainer_guide.md:5]: "Internal" as "internal meta" wiki?

File: maintainer_guide.md
Comment (pmeier) [inline maintainer_guide.md:9]: I'm never sure how we should call core: torch or PyTorch?

File: maintainer_guide.md
Comment (pmeier) [inline maintainer_guide.md:11]: Can we link the policies here?

File: maintainer_guide.md
Comment (pmeier) [inline maintainer_guide.md:19]: Are you referring to the situation discussed offline, i.e. defining public function in a private namespace and expose them in a public one? If so, can we clarify the text? The "exposed / aliased" implies for me that if we expose `_foo` in a public `__init__.py`, it can be considered public. Was that your intention?

File: maintainer_guide.md
Comment (pmeier) [inline maintainer_guide.md:27]: Since this guide is not aimed purely at Meta employees, we should clarify this. Probably just preface this with "For Meta employees:"

File: maintainer_guide.md
Comment (pmeier) [inline maintainer_guide.md:26]:
```suggestion
change. We should be careful about models running in production and relying on
```

File: maintainer_guide.md
Comment (pmeier) [inline maintainer_guide.md:38]:
```suggestion
The rest of the folders (build, android, ios, etc.) are private and have no BC
```

Hammer home that we actually mean all other folders

**Review by pmeier — COMMENTED** [review]
Mostly minor comments, thanks Nicolas. Otherwise I agree to have something like that. Let's pull in @vfdev-5 too as the three of us are the only active maintainers at this point.

File: maintainer_guide.md
Comment (NicolasHug) [reply maintainer_guide.md:9]: over here it's usually "torch core" or "core", but "PyTorch" isn't ambiguous either. Happy with whatever you prefer

File: maintainer_guide.md
Comment (NicolasHug) [reply maintainer_guide.md:19]: Happy to clarify, I tried "exposed / aliased **as public**" but maybe it's not clear enough. Perhaps

> exposed / aliased **as public** (i.e. without a leading underscore) in a public `__init__.py` file

?

Comment (NicolasHug) [reply maintainer_guide.md:19]: Maybe a simpler way is to just say:

> Any object exposed in a public `__init__.py` is public unless it has a leading underscore in its immediate name

?

Comment (pmeier) [reply maintainer_guide.md:19]: Yes, I like that a lot better

File: maintainer_guide.md
Comment (pmeier) [reply maintainer_guide.md:9]: In that case, I'm in favor of "torch" since with that we don't have a disconnect in the naming scheme when talking about "torchvision". (I really hate that in `conda install pytorch torchvision -c pytorch`)

File: maintainer_guide.md
Comment (NicolasHug) [reply maintainer_guide.md:21]: So it's just a special case of what we have above:

> - objects in a public file that don't have a leading underscore are public

Do you think it's worth re-iterating that with `__init__.py` ?

File: maintainer_guide.md
Comment (vfdev-5) [inline maintainer_guide.md:49]: I would assume that this document is not intended to users. However this part and other part about C++ being private, it worth exposing to the users.

File: maintainer_guide.md
Comment (pmeier) [reply maintainer_guide.md:21]: Technically yes, since `__init__.py` itself is not a public file given its underscores. But since have stated above that our definition might technically not be perfect and this should be common knowledge for Python devs, I'm ok to merge the two.

File: maintainer_guide.md
Comment (NicolasHug) [reply maintainer_guide.md:9]: Sorry to be pedantic but  "torch" may refer to the original Lua torch ? https://en.wikipedia.org/wiki/Torch_(machine_learning)

File: maintainer_guide.md
Comment (NicolasHug) [reply maintainer_guide.md:48]: I agree...
At the same time, I don't know where we could expose it properly. Pytorch typically doesn't have that in the online docs (I think?), it's rather in the Wiki which most users wouldn't unless they're actively searching for it.

Perhaps changing the title from `maintiner_guie` to `policies` or something like that? Happy to hear any other suggestion

**Comment by NicolasHug** [comment]
@pmeier @vfdev-5 any additional concern / comment I can address before merging?

File: maintainer_guide.md
Comment (pmeier) [inline maintainer_guide.md:31]:
```suggestion
projects should *not* rely on it (it has happened in the past and is now
```

File: maintainer_guide.md
Comment (pmeier) [inline maintainer_guide.md:32]:
```suggestion
programmatically impossible).
```

**Review by pmeier — APPROVED** [review]
Two nits, but otherwise LGTM! Thanks Nicolas!

**Review by vfdev-5 — APPROVED** [review]
LGTM as well, thanks @NicolasHug

**Pull request merged**
